### DISCUSSION TITLE
How to configure audio devices with ovos?

### DISCUSSION THREAD

**Discussion by dangarbri** [OpenVoiceOS/ovos-buildroot#267] [Q&A]
I'm trying to run ovos on a raspberry pi 4. I'm using an IQaudio codec zero for the audio i/o. It needs some configuration to use. I'm able to set it up and get it working on raspbian by editing `/boot/firmware/config.txt` and `/etc/rc.local` to enable the right drivers and startup scripts.

Is it possible to do this on the raspOVOS images? I didn't see a `config.txt` when I installed the img to an sdcard.

**Comment by builderjer** [comment]
Thank you for testing the raspOVOS image.  We would love to help you get this working, and maybe even add it into our [ovos-i2csound](https://github.com/OpenVoiceOS/ovos-i2csound) for auto detection and setup.  There are a few things I need to know and a few statements.

> Is it possible to do this on the raspOVOS images?

raspOVOS is based directly on raspbian and the images are built using the official build tool, [pi-gen](https://github.com/RPi-Distro/pi-gen), so with few exceptions, what you can do there, you can do here.

> I didn't see a config.txt when I installed the img to an sdcard.

raspbian has `config.text` in a strange place before the first boot.  This allows the image to place it in different files depending on the board used.
After the first boot of raspOVOS, the file should be at `/boot/firmware/config.txt`

Now to your sound card.

It looks like this device has an EEPROM so i2cdetect should work.
Can you post the output of a few commands
 `i2cdetect -y 1`
`arecord -l`
`aplay -l`

I hope we can get this working, and looks like a good candidate for auto configuration.

**Reply by dangarbri** [reply]
Thanks! I'll try it again tomorrow and report back.

**Reply by dangarbri** [reply]
```
(.venv) ovos@raspOvos:~ $ i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- UU -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
(.venv) ovos@raspOvos:~ $ arecord -l
**** List of CAPTURE Hardware Devices ****
card 2: Zero [RPi Codec Zero], device 0: Raspberry Pi Codec Zero HiFi da7213-hifi-0 [Raspberry Pi Codec Zero HiFi da7213-hifi-0]
  Subdevices: 0/1
  Subdevice #0: subdevice #0
(.venv) ovos@raspOvos:~ $ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: b1 [bcm2835 HDMI 1], device 0: bcm2835 HDMI 1 [bcm2835 HDMI 1]
  Subdevices: 4/4
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
card 0: b1 [bcm2835 HDMI 1], device 1: bcm2835 HDMI 1 [bcm2835 HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]
  Subdevices: 4/4
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
card 2: Zero [RPi Codec Zero], device 0: Raspberry Pi Codec Zero HiFi da7213-hifi-0 [Raspberry Pi Codec Zero HiFi da7213-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
(.venv) ovos@raspOvos:~ $ 
```

The Zero is detected, but I can't get it to play sound.
I know the speaker is wired in correctly because it works with raspbian desktop, like I've said before.
This is an issue with setting the speaker as the default audio device.

**Reply by dangarbri** [reply]
Per the official instructions: https://www.raspberrypi.com/documentation/accessories/audio.html#hardware-versions

config.txt
```
(.venv) ovos@raspOvos:~ $ cat /boot/firmware/config.txt
# For more options and information see
# http://rptl.io/configtxt
# Some settings may impact device functionality. See link above for details

# Uncomment some or all of these to enable the optional hardware interfaces
dtparam=i2c_arm=on
dtparam=i2s=on
dtparam=spi=on

# Enable audio (loads snd_bcm2835)
dtparam=audio=on

# Additional overlays and parameters are documented
# /boot/firmware/overlays/README

# Automatically load overlays for detected cameras
camera_auto_detect=1

# Automatically load overlays for detected DSI displays
display_auto_detect=1

# Automatically load initramfs files, if found
auto_initramfs=1

# Enable DRM VC4 V3D driver
dtoverlay=vc4-fkms-v3d
max_framebuffers=2

# Don't have the firmware create an initial video= setting in cmdline.txt.
# Use the kernel's default instead.
disable_fw_kms_setup=1

# Run in 64-bit mode
arm_64bit=1

# Disable compensation for displays with overscan
disable_overscan=1

# Run as fast as firmware / board allows
arm_boost=1

[cm4]
# Enable host mode on the 2711 built-in XHCI USB controller.
# This line should be removed if the legacy DWC2 controller is required
# (e.g. for USB device mode) or if USB support is not required.
otg_mode=1

[all]


disable_splash=1

[pi3]

force_turbo=1
boot_delay=1

[all]

dtoverlay=miniuart-bt

dtoverlay=i2s-mmap=on
gpu_mem=256

dtoverlay=rpi-codeczero
```
/etc/rc.local
```
(.venv) ovos@raspOvos:~ $ cat /etc/rc.local
#!/bin/sh -e
#
# rc.local
#
# This script is executed at the end of each multiuser runlevel.
# Make sure that the script will "exit 0" on success or any other
# value on error.
#
# In order to enable or disable this script just change the execution
# bits.
#
# By default this script does nothing.

# Print the IP address
_IP=$(hostname -I) || true
if [ "$_IP" ]; then
  printf "My IP address is %s\n" "$_IP"
fi

sudo alsactl restore -f /home/ovos/Pi-Codec/Codec_Zero_OnboardMIC_record_and_SPK_playback.state

exit 0
```

The alsa state file exists:
```
(.venv) ovos@raspOvos:~ $ ls /home/ovos/Pi-Codec/Codec_Zero_OnboardMIC_record_and_SPK_playback.state
/home/ovos/Pi-Codec/Codec_Zero_OnboardMIC_record_and_SPK_playback.state
```

asoundrc
```
(.venv) ovos@raspOvos:~ $ cat ~/.asoundrc
pcm.!default {
  type hw
  card Zero
}
```

**Reply by dangarbri** [reply]
One more update, I commented out `dtparam=audio=on` in `config.txt`

Now my only audio device is the IQAudio, and it doesn't work...

```
(.venv) ovos@raspOvos:~ $ i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- UU -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
(.venv) ovos@raspOvos:~ $ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: Zero [RPi Codec Zero], device 0: Raspberry Pi Codec Zero HiFi da7213-hifi-0 [Raspberry Pi Codec Zero HiFi da7213-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
(.venv) ovos@raspOvos:~ $ arecord -l
**** List of CAPTURE Hardware Devices ****
card 0: Zero [RPi Codec Zero], device 0: Raspberry Pi Codec Zero HiFi da7213-hifi-0 [Raspberry Pi Codec Zero HiFi da7213-hifi-0]
  Subdevices: 0/1
  Subdevice #0: subdevice #0
(.venv) ovos@raspOvos:~ $ cat ~/.asoundrc
pcm.!default {
  type hw
  card Zero
}
(.venv) ovos@raspOvos:~ $ speaker-test

speaker-test 1.2.8

Playback device is default
Stream parameters are 48000Hz, S16_LE, 1 channels
Using 16 octaves of pink noise
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 192 to 2097152
Period size range from 64 to 699051
Using max buffer size 2097152
Periods = 4
was set period_size = 524288
was set buffer_size = 2097152
 0 - Front Left
Time per period = 10.964373
 0 - Front Left
```

`speaker-test` thinks its working, which is farther than I got before. The speaker makes like a clicking sound when speaker-test starts and then goes silent, no white noise like it should. I had a similar issue on raspbian headless, but on raspbian headless speaker-test wouldn't start at all for the device.

**Comment by j1nx** [comment]
If they cloned the original Raspberry Device correctly the DtOverlay is actually saved within and loaded from the EEPROM and it should automatically be configured. If not;
https://www.raspberrypi.com/documentation/accessories/audio.html#hardware-versions

All required information is accesable from that link, including the link to the ALSA state files at the rapberry pi github repositories to get going quickly.

**Reply by dangarbri** [reply]
Using [rpi-imager to install raspberrypi os with the desktop](https://www.raspberrypi.com/software/) works well.
I was able to set this up and install ovos in docker via the [ovos-installer](https://openvoiceos.github.io/community-docs/044-ovos_installer/).

But I wasn't able to run raspOVOS, which is what my question was really about. Sorry, probably a poor title on my part.

p.s.
I have a separate issue with that audio documentation though, I opened up on issue on those docs page. It works fine with raspbian desktop, but not headless raspbian with cli only. But I don't think that's relevant to my question here.

**Reply by builderjer** [reply]
What was the issue with raspOVOS? Would it not boot? Or something else.

What is your hardware and I will try and replicate

**Reply by dangarbri** [reply]
I might be mixing up raspOVOS and the buildroot image.

What happened was I installed the image, and when I powered it on connected to a monitor I saw the OVOS logo, but nothing happened, it just stayed there.

I'm getting the raspOVOS image to try again now.

The hardware is:
- [Raspberry PI 4B w/4gb memory](https://www.adafruit.com/product/4296)
- [IQAudio Codec Zero](https://www.pishop.us/product/iqaudio-codec-zero/) - This was rebranded to a [Raspberry PI Codec Zero](https://www.raspberrypi.com/products/codec-zero/) with a green board, I have the original IQAudio black board.
- [Speaker](https://www.adafruit.com/product/1313)

**Reply by builderjer** [reply]
I do not have your sound card. But if it works on raspbian,  it will work with raspOVOS. 

If the gui image does not boot, please let me know. 

I will test it with a pi4 and respeaker that I have. It will at least test the image. We can work on detection after that. 

If you did try the buildroot image, I think it is old. There has been A LOT going on there BTS. A new one should be available soon. No time-line though. Like I said, a lot going on there

**Reply by dangarbri** [reply]
Ok, the raspOVOS image I just tried doesn't seem to work
I flashed the sdcard with rpi-imager, I did set custom user settings with rpi-imager, I don't know if that caused problems.

It powered on, said it was generating ssh keys, then rebooted.
Now when it boots, I see the startup kernel output, and then the hdmi goes out to no signal.
I used the image from [here](https://ovosimages.ziggyai.online/raspbian/newest/) which the docs do say are unstable.

**Reply by dangarbri** [reply]
I re-flashed without trying to set settings with rpi-imager and it boots. I'm not sure the speaker will work with this setup though. With raspbian I had to use the gui to enable the speaker as the default device. Setting it with `.asoundrc` doesn't work, even though thats what the docs say to do.

**Reply by dangarbri** [reply]
How do I get a terminal from here? I have a usb keyboard, but no usb mouse.

![IMG_20240506_195521.jpg](https://github.com/OpenVoiceOS/ovos-buildroot/assets/6597039/7fcc54a6-59e4-4091-aaa0-5c213eb31b5c)

**Reply by dangarbri** [reply]
Got a terminal with alt+sysrq+e to kill the gui process.

**Reply by builderjer** [reply]
Yes, changing settings with the imager has been known to cause issues. I think I have the username hard coded in some places, especially giving ovos root access. Can you try again without modified settings?

And like I said, I will test it again as soon as I can.

**Reply by builderjer** [reply]
From the image above, if you attach a mouse and pull down from the top. It should open a menu. Click the wifi button, and you should be able to go from there. Choose on Mobile setup.

**Reply by dangarbri** [reply]
I don't have a usb mouse :(
I can't start blueooth with raspovos?

`systemctl start bluetooth`
`systemctl status bluetooth`
`Bluetooth service was skipped buecause of an unmet condition check (ConditionPathIsDirectory=/sys/class/bluetooth)`

**Reply by dangarbri** [reply]
Needed to start the `hciuart` service

**Reply by dangarbri** [reply]
Ok, I have raspovos running, but there's no sound :( This is where I got stuck with the raspbian headless image.
https://github.com/raspberrypi/documentation/issues/3593#issuecomment-2065313967

**Reply by builderjer** [reply]
you could disable ovos-i2csound and set it up manually.
`sudo systemctl disable i2csound.service`
Then it should not interfere with your setup.  I will see what I can do to see if there is another way to detect this device

**Reply by dangarbri** [reply]
I'll try this when I get home

**Reply by dangarbri** [reply]
I disabled i2csound but it's still not working. alsamixer shows my device as the default device, but there's also pulseaudio. Maybe they're not working together.

![image](https://github.com/OpenVoiceOS/ovos-buildroot/assets/6597039/4eaddf3e-95fa-4ae2-95c9-2804ca9f2b59)
![image](https://github.com/OpenVoiceOS/ovos-buildroot/assets/6597039/4b496421-e8d6-4f53-a8ee-73640e58b2dd)
![image](https://github.com/OpenVoiceOS/ovos-buildroot/assets/6597039/296fb8b7-efba-4e18-a4f8-9c22184c72e4)

**Reply by dangarbri** [reply]
`sudo systemctl disable i2csound.service` doesn't disable i2csound. It still runs on startup.

```
○ i2csound.service - I2C Sound configuration service
     Loaded: loaded (/etc/systemd/system/i2csound.service; static)
     Active: inactive (dead) since Tue 2024-05-07 20:03:10 MDT; 7min ago
    Process: 399 ExecStart=/usr/libexec/ovos-i2csound (code=exited, status=0/SUCCESS)
   Main PID: 399 (code=exited, status=0/SUCCESS)
        CPU: 366ms

May 07 20:03:10 raspOvos ovos-i2csound[621]: Invalid card number 'wm8960soundcard'.
May 07 20:03:10 raspOvos ovos-i2csound[622]: Invalid card number 'wm8960soundcard'.
May 07 20:03:10 raspOvos ovos-i2csound[623]: Invalid card number 'wm8960soundcard'.
May 07 20:03:10 raspOvos ovos-i2csound[624]: Invalid card number 'wm8960soundcard'.
May 07 20:03:10 raspOvos ovos-i2csound[625]: Invalid card number 'wm8960soundcard'.
May 07 20:03:10 raspOvos ovos-i2csound[626]: Invalid card number 'wm8960soundcard'.
May 07 20:03:10 raspOvos ovos-i2csound[627]: Invalid card number 'wm8960soundcard'.
May 07 20:03:10 raspOvos ovos-i2csound[399]: Configuring button
May 07 20:03:10 raspOvos systemd[1]: i2csound.service: Deactivated successfully.
May 07 20:03:10 raspOvos systemd[1]: Finished i2csound.service - I2C Sound configuration service.
```

Apparently the "static" state there means it can't be disabled. So instead I moved `i2csound.service` out of the `/etc/systemd/system` directory to disable it. But still no improvement.

edited to add full systemctl status output

**Comment by j1nx** [comment]
@builderjer This; https://github.com/raspberrypi/linux/blob/fd553e002cb908f0dd681af0b3ddb62965e0cc41/arch/arm/boot/dts/overlays/iqaudio-codec-overlay.dts#L25)

MIght be problematic for i2csound.

**Reply by builderjer** [reply]
I would say so.  It would register as a wm8960.  Maybe there is another check that can distinguish it from others.  I will look some more

**Reply by j1nx** [reply]
I do have an idea though.

as the dtoverlay is saved within the EEPROM, by the time the bootprocess reaches i2csound it might already be loaded up.

so when it sees "UU" at that adres without it being explicitely loaded from config.txt consider it this card. If not consider it a normal rpi-proto card (they do however use the exact same codec / dtoverlay as the rpi-proto name got renamed to rpi-codec)

https://github.com/OpenVoiceOS/ovos-i2csound/blob/dev/ovos-i2csound#L142

**Reply by dangarbri** [reply]
Is it possible that other devices would show up as UU here? What is this UU? Some device ID?

**Reply by goldyfruit** [reply]
> Is it possible that other devices would show up as UU here? What is this UU? Some device ID?

Yes it is, `UU` means that the device is handled by a kernel module/consumed.

**Reply by dangarbri** [reply]
I don't think its safe to assume its this card specifically if other devices would look the same.


Is i2csound meant to configure audio devices not managed by the kernel? Would that mean if you see this UU then you would want i2csound to do skip configuring them.

**Reply by JarbasAl** [reply]
i2csound runs on boot and detects the hardware connected, configuring the audio automatically

this allows the prebuilt images to work out of the box with several hardware, and even to just reuse the same USB drive for testing

i suggest we move this part of the discussion to a new issue in i2csound to explicitly track support for this hat

we absolutely want to avoid false detections too! this means some devices simply cant be configured automatically since we do not want to guess

**Comment by wita-sec** [comment]
I also had a couple of issues with OVOS audio (Mic and Speaker) on my RPi4. 
Maybe pavucontrol and pulsemixer https://github.com/GeorgeFilipkin/pulsemixer can shed some light on your audio configuration.
For me these tools were very helpful to get audio running.
